src/geom/SkyWcs.cc

```diff
@@ -0,0 +1,249 @@
+    double xmult = flipX ? -1.0 : 1.0;
+    cdMatrix(0, 0) = std::cos(orientRad) * scaleDeg * xmult;
+    cdMatrix(0, 1) = std::sin(orientRad) * scaleDeg;
+    cdMatrix(1, 0) = -std::sin(orientRad) * scaleDeg * xmult;
+    cdMatrix(1, 1) = std::cos(orientRad) * scaleDeg;
+    return cdMatrix;
+}
+
+SkyWcs::SkyWcs(Point2D const& crpix, SpherePoint const& crval, Eigen::Matrix2d const& cdMatrix)
+        : SkyWcs(detail::readLsstSkyWcs(*detail::makeTanWcsMetadata(crpix, crval, cdMatrix))) {}
+
+SkyWcs::SkyWcs(daf::base::PropertyList& metadata) : SkyWcs(detail::readLsstSkyWcs(metadata)) {}
+
+SkyWcs::SkyWcs(ast::FrameSet const& frameSet)
+        : Transform<Point2Endpoint, SpherePointEndpoint>(_checkFrameSet(frameSet)) {}
+
+Angle SkyWcs::getPixelScale(Point2D const& pixel) const {
+    // Compute pixVec containing the pixel position of three corners of the pixel
+    // (It would be slightly more accurate to measure across the pixel center,
+    // but that would require one additional pixel-to-sky transformation)
+    std::vector<Point2D> pixVec;
+    pixVec.push_back(pixel - Extent2D(0.5, 0.5));   // lower left corner
+    pixVec.push_back(pixel + Extent2D(0.5, -0.5));  // lower right corner
+    pixVec.push_back(pixel + Extent2D(-0.5, 0.5));  // upper left corner
+
+    auto skyVec = tranForward(pixVec);
+
+    // Work in 3-space to avoid RA wrapping and pole issues.
+    auto skyLL = skyVec[0].getVector();
+    auto skyDx = skyVec[1].getVector() - skyLL;
+    auto skyDy = skyVec[2].getVector() - skyLL;
+
+    // Compute |cross product| = area of parallelogram with sides dx,dy, in radians^2
+    // This is slightly incorrect -- it's making the small-angle
+    // approximation, taking the distance *through* the unit sphere
+    // rather than over its surface, but the difference should be negligible for pixels of normal size
+    double pixAreaRadSq = std::sqrt(square(skyDx[1] * skyDy[2] - skyDx[2] * skyDy[1]) +
+                                    square(skyDx[2] * skyDy[0] - skyDx[0] * skyDy[2]) +
+                                    square(skyDx[0] * skyDy[1] - skyDx[1] * skyDy[0]));
+    return std::sqrt(pixAreaRadSq) * radians;
+}
+
+Point2D SkyWcs::getPixelOrigin() const { return this->tranInverse(getSkyOrigin()); }
+
+SpherePoint SkyWcs::getSkyOrigin() const {
+    // CRVAL is stored as the SkyRef property of the sky frame (the current frame of the SkyWcs)
+    auto const skyFrame = detail::getSkyFrame(*_frameSet, ast::FrameSet::CURRENT, false);
+    auto const crvalRad = skyFrame->getSkyRef();
+    return SpherePoint(crvalRad.data());
+}
+
+Eigen::Matrix2d SkyWcs::getCdMatrix() const {
+    // Use a copy because looking up the IWC frame alters the FrameSet
+    auto frameSet = getFrameSet()->copy();
+    auto prototype = ast::Frame(2, "Domain=IWC");
+    auto iwcFrameSet = frameSet->findFrame(prototype);
+    if (!iwcFrameSet) {
+        throw LSST_EXCEPT(lsst::pex::exceptions::RuntimeError, "IWC frame not found");
```

Comment:
But the general comment is that this submission tends to call everything a `RuntimeError`.  It's important to use the appropriate exception type, because the *type* of the exception is important for categorizing errors, both manually and in the code.